In [20]:
import numpy as np
import pandas as pd
import matplotlib.pylab as plt
plt.style.use('ggplot')

from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
#import logistic regression
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [2]:
df=pd.read_csv('creditcard.csv')

In [ ]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [ ]:
# 1. Total Number of Null Values
df.isnull().sum() 


Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [ ]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

Dataset is unbalanced
0: normal transation
1: Fradulent Transaction

In [5]:
#Seperate the data for analysis
legit = df[df['Class']==0]
fraud = df[df['Class']==1]


In [ ]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [ ]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [ ]:
fraud.Amount.describe()
#20.25

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [3]:
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

#Number of Fradulent Transcations --> 492 
#Number of Legitment Transcations --> 284,315

In [6]:
#Under-Sampling, Build a sample dataset containing similar distribution of normal transactions and Fraudulent Transactions
legit_sample = legit.sample(n=492)


In [7]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [8]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
193383  130114.0 -0.546853  0.724351 -0.688233 -0.489711  1.239061 -1.069653   
31479    36382.0  1.136452  0.008456 -0.057826  0.859583 -0.159075  0.139526   
259763  159263.0  1.999871 -0.103139 -1.388265  0.160581  0.240692 -0.592323   
188596  128044.0 -0.208404 -0.433350 -1.672835 -4.786898  1.977587  3.327441   
39496    39861.0 -1.464009  0.160811  1.673052 -0.906483 -0.661852 -0.202896   

              V7        V8        V9  ...       V21       V22       V23  \
193383  0.909243 -0.331679  0.235357  ...  0.168384  0.642288 -0.211599   
31479  -0.393983  0.316106  0.634696  ... -0.144604 -0.266602 -0.095577   
259763  0.101517 -0.177957  0.292479  ...  0.323884  1.007869  0.006103   
188596 -1.138778 -1.915787 -0.509803  ... -1.859007 -1.157190 -0.034466   
39496  -0.024260  0.616975  0.544253  ... -0.109280 -0.415542  0.050007   

             V24       V25       V26       V27       V28  Amount  Class  
193383 -0.803211  0.376324 -0.095482 -0.316931 -0.015896    9.99      0  
31479  -0.408181  0.424593  0.449850  0.001202  0.022155   12.31      0  
259763  0.814137  0.246050 -0.161136 -0.025531 -0.056119   19.95      0  
188596  0.633227  0.978939 -0.590704  0.051091  0.231493   79.98      0  
39496  -0.057049  0.040780  0.824893  0.117338  0.034262   93.82      0  

[5 rows x 31 columns]

In [9]:
new_dataset.value_counts("Class")

Class
0    492
1    492
dtype: int64

In [11]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92848.750000 -0.020006 -0.032785  0.046647 -0.000325  0.025548   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.119880  0.079288 -0.075360 -0.087300  ...  0.001219 -0.032586   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23      V24       V25       V26       V27       V28  \
Class                                                                        
0     -0.003088 -0.013882  0.00884 -0.003126 -0.004096  0.004483 -0.019030   
1      0.014049 -0.040308 -0.10513  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       93.557175  
1      122.211321  

[2 rows x 30 columns]

# Splitting the Data into Features and Targets

In [12]:
X= new_dataset.drop(columns="Class",axis=1)
Y= new_dataset["Class"]

In [13]:
X.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount'],
      dtype='object')

In [14]:
print(Y)

193383    0
31479     0
259763    0
188596    0
39496     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


In [15]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2, stratify=Y, random_state=2)
#20% data in test set, 80% data in training set

In [16]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


In [21]:
model=LogisticRegression()

In [22]:
#training the Logistic Regression Model with Training Data
model.fit(X_train,Y_train)

LogisticRegression()

In [23]:
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [24]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9339263024142312


In [25]:
#Accuracy On Testing Data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [26]:
print('Accuracy on Test data : ', test_data_accuracy)

Accuracy on Test data :  0.8781725888324873
